### PR TITLE
Upgrade okhttp to 4.10.0

### PR DESCRIPTION
**Pull request by wendigo**
NO RELEASE NOTES

### CODE REVIEW

**Review by electrum — COMMENTED** [review]
We intentionally stick with the 3.x branch because it's written in Java. What is the motivation to upgrade?

If we upgrade to the Kotlin version, we need to update the shading rules for the JDBC driver to include the additional packages.

File: plugin/trino-druid/pom.xml [hunk 1]
```diff
@@ -77,6 +77,12 @@
             </exclusions>
         </dependency>
 
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
+            <scope>compile</scope>
```
Comment (electrum) [inline plugin/trino-druid/pom.xml:83]: `compile` is default scope and should never be listed explicitly

File: client/trino-client/src/main/java/io/trino/client/OkHttpUtil.java [hunk 2]
```diff
@@ -226,7 +226,7 @@ public static void setupSsl(
             sslContext.init(keyManagers, new TrustManager[] {trustManager}, null);
 
             clientBuilder.sslSocketFactory(sslContext.getSocketFactory(), trustManager);
-            clientBuilder.hostnameVerifier(LegacyHostnameVerifier.INSTANCE);
+            clientBuilder.hostnameVerifier(OkHostnameVerifier.INSTANCE);
```
Comment (electrum) [inline client/trino-client/src/main/java/io/trino/client/OkHttpUtil.java:229]: This is a behavioral change. The legacy verifier allows certificates which use the CN rather than SubjectAltName. While this violates modern TLS specifications, it is common in internal environments, and even our own unit tests (unless we've since fixed that).

Comment (wendigo) [reply client/trino-client/src/main/java/io/trino/client/OkHttpUtil.java:229]: I've added a legacy hostname verifier with changes required for it to work (some of the code that this class is depending on was removed)

**Comment by findepi** [comment]
> We intentionally stick with the 3.x branch because it's written in Java. What is the motivation to upgrade?

The Kotlin version might include https://github.com/square/okio/pull/818, a fix for potential deadlock in the library.

**Comment by wendigo** [comment]
@electrum @findepi ptal

**Comment by wendigo** [comment]
What's the veridct here @findepi @electrum ?

File: plugin/trino-druid/pom.xml [hunk 3]
```diff
@@ -77,6 +77,11 @@
             </exclusions>
         </dependency>
 
+        <dependency>
```
Comment (electrum) [inline plugin/trino-druid/pom.xml:80]: This dependency shouldn't' be required at all for this module, since the classes aren't referenced here.

File: client/trino-jdbc/pom.xml [hunk 4]
```diff
@@ -353,6 +357,14 @@
                                     <pattern>net.jodah.failsafe</pattern>
                                     <shadedPattern>${shadeBase}.net.jodah.failsafe</shadedPattern>
                                 </relocation>
+                                <relocation>
```
Comment (electrum) [inline client/trino-jdbc/pom.xml:360]: Make all these dependencies `<optional>true</optional>` like we do for `jsr305` and other annotation-only dependencies. No need to include them in the final driver JAR.

File: client/trino-client/src/main/java/okhttp3/internal/tls/LegacyHostnameVerifier.java [hunk 5]
```diff
@@ -19,51 +23,646 @@
...
-import static okhttp3.internal.tls.OkHostnameVerifier.allSubjectAltNames;
-
-public class LegacyHostnameVerifier
+/**
+ * A HostnameVerifier that accepts hostnames in both the standard location (subjectAltName)
+ * and the obsolete CNAME location, which deprecated in RFC 2818 (May 2000).
+ * <p>
+ * Source: https://gist.github.com/swankjesse/461a53ee3078bdaa574e3d4e48aeaf07
+ */
+public final class LegacyHostnameVerifier
```
Comment (electrum) [inline client/trino-client/src/main/java/okhttp3/internal/tls/LegacyHostnameVerifier.java:41]: Rather than copy this whole class, we can leave the original implementation and simply copy the last version of [`DistinguishedNameParser`](https://raw.githubusercontent.com/square/okhttp/fc7ac8ae0b30d219cbd884b2af458f19c170c5fb/okhttp/src/main/java/okhttp3/internal/tls/DistinguishedNameParser.java) from OkHttp, which I confirmed is identical the one we are already using. This way, in the normal path, we are using the TLS verification from OkHttp, rather than relying on an implementation from a four year old gist (that is neither tested nor updated with any security fixes). Might also need to copy the  `Util.verifyAsIpAddress()` method, but the others should still be there.

File: plugin/trino-http-event-listener/pom.xml [hunk 6]
```diff
@@ -91,7 +91,7 @@
         <dependency>
```
Comment (electrum) [inline plugin/trino-http-event-listener/pom.xml:91]: We should be able to remove this dependency now. It shouldn't have been added in the first place.

**Review by electrum — COMMENTED** [review]
Upgrading makes sense. We can't stay on the old version forever. I'll take another look at the hostname verifier before approving, since this is security critical code.

File: client/trino-jdbc/pom.xml (hunk 4 above)
Comment (electrum) [reply client/trino-jdbc/pom.xml:360]: Actually, exclude all the annotations in the `<dependency>` section, like we do for the annotations for Guava in this module.

**Comment by wendigo** [comment]
Thx @electrum

**Comment by mosabua** [comment]
👋 @wendigo .. could you pick this up again and work with @electrum to get it merged?

**Comment by wendigo** [comment]
@mosabua yes, if we agree that we need it (and I guess that we do)

File: plugin/trino-druid/pom.xml (hunk 3 above)
Comment (wendigo) [reply plugin/trino-druid/pom.xml:80]: They are in the test (@Language annotation)

File: plugin/trino-http-event-listener/pom.xml (hunk 6 above)
Comment (wendigo) [reply plugin/trino-http-event-listener/pom.xml:91]: Actually maven complains that it's used by the test (which isn't the case and I don't understand why's that):

```
[INFO] --- maven-dependency-plugin:3.3.0:analyze-only (default) @ trino-http-event-listener ---
[WARN] Used undeclared dependencies found:
[WARN]    com.squareup.okio:okio-jvm:jar:3.0.0:test
```

**Comment by wendigo** [comment]
@electrum ptal

**Comment by wendigo** [comment]
I've checked that trino-jdbc is shaded properly

**Comment by wendigo** [comment]
PTs have failed because the executable jar is too big in size (number of entries exceeds 64K). We've already encountered this problem in the past and I have working solution to this (moving to executable zip64 format). I'll tackle that next week. 

Cc: @nineinchnick @electrum @findepi

File: client/trino-client/src/main/java/okhttp3/internal/tls/DistinguishedNameParser.java [hunk 7]
```diff
@@ -0,0 +1,426 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package okhttp3.internal.tls;
+
+import javax.security.auth.x500.X500Principal;
+
+// Formatted copy of https://raw.githubusercontent.com/square/okhttp/fc7ac8ae0b30d219cbd884b2af458f19c170c5fb/okhttp/src/main/java/okhttp3/internal/tls/DistinguishedNameParser.java
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/okhttp3/internal/tls/DistinguishedNameParser.java:18]: Can you mention the reason to keep the copy here, and why you've chosen this particular commit? I tried figuring this out myself but it's not trivial. When I open https://github.com/square/okhttp/commit/fc7ac8ae0b30d219cbd884b2af458f19c170c5fb it is an unrelated change, and closest tags don't help:
obraz

Is this the last commit from the 3.x version we're using?

Comment (wendigo) [reply client/trino-client/src/main/java/okhttp3/internal/tls/DistinguishedNameParser.java:18]: Yes @nineinchnick. David pointed me to this one

File: client/trino-client/src/main/java/okhttp3/internal/tls/LegacyHostnameVerifier.java [hunk 8]
```diff
@@ -66,4 +75,129 @@ public boolean verify(String host, SSLSession session)
             return false;
         }
     }
+
+    private boolean verifyAsIpAddress(String host)
+    {
+        return VERIFY_AS_IP_ADDRESS.matcher(host).matches();
+    }
+
+    private boolean verifyHostname(String hostName, String pattern)
```
Comment (nineinchnick) [inline client/trino-client/src/main/java/okhttp3/internal/tls/LegacyHostnameVerifier.java:84]: If this is also a copy from OkHttp, can you add a comment, similar to the one in `DistinguishedNameParser.java`?

File: plugin/trino-pinot/pom.xml [hunk 9]
```diff
@@ -500,6 +500,12 @@
             <scope>runtime</scope>
         </dependency>
 
+        <dependency>
+            <groupId>org.jetbrains</groupId>
+            <artifactId>annotations</artifactId>
+            <scope>runtime</scope>
```
Comment (nineinchnick) [inline plugin/trino-pinot/pom.xml:506]: I assume this change is related, but I'm not sure how.

Comment (wendigo) [reply plugin/trino-pinot/pom.xml:506]: This dependency is now transitively used.

**Comment by wendigo** [comment]
@nineinchnick I've switched to forked (and now released under io.starburst) executable-maven-plugin

**Comment by wendigo** [comment]
Woah. All tests have passed

**Comment by mosabua** [comment]
> Woah. All tests have passed

So ... did we want to merge this for 404 maybe?

File: client/trino-cli/pom.xml [hunk 10]
```diff
@@ -159,7 +159,7 @@
             </plugin>
 
             <plugin>
-                <groupId>org.skife.maven</groupId>
```
[the remark this answers is not in the document]
Comment (wendigo) [reply client/trino-cli/pom.xml:162]: Repo seemed dead to me. That's why I've rewritten the plugin

[the remark this answers is not in the document]
Comment (martint) [reply client/trino-cli/pom.xml:162]: Stable doesn't mean dead. It's a tiny project, so it's expected to not be very active.

[the remark this answers is not in the document]
Comment (wendigo) [reply client/trino-cli/pom.xml:162]: Is there aby reason why proceeding with the fork would be a blocker?

[the remark this answers is not in the document]
Comment (martint) [reply client/trino-cli/pom.xml:162]: We don't want to maintain a fork if the original project is still alive.

**Comment by wendigo** [comment]
PTAL @kokosing @findepi

**Review by kokosing — COMMENTED** [review]
Please make it clear what code is copy pasted and are the modification you have done to it. Possibly we may want to add copied code as separate commit. Should we do the same for copied methods and store them in dedicated class?

@electrum Are you willing to take a look at `LegacyHostnameVerifier`? From what I hear all your concerns are addressed.

**Comment by kokosing** [comment]
Please also add a test to check checksum to prevent accidental modification of copy-pasted code.

**Comment by electrum** [comment]
We use OkHttp for the Trino client (CLI, JDBC driver). Why are we touching Kafka, Avro, etc.? Can you scope this just to the client changes? If we have other usages of OkHttp, let's address those separately.

**Comment by wendigo** [comment]
@electrum okhttp has some shared dependencies with wiremock. It's basically a dependency hell. I've been able to update successfuly a bunch of those dependencies and they seem to work with each other.

File: pom.xml [hunk 11]
```diff
@@ -198,6 +198,13 @@
         <module>testing/trino-tests</module>
     </modules>
 
+    <repositories>
```
Comment (electrum) [inline pom.xml:201]: We don't want to add a new repository definition in the root POM, as it will be used for all modules. Ideally, we wouldn't have this at all, but at least it's currently limited to a few modules.

**Comment by electrum** [comment]
Hmm, is it possible to upgrade the Kafka stuff first, without upgrading OkHttp? This is a large change that's hard to understand and review.

**Comment by wendigo** [comment]
I can give it a try @electrum

**Comment by kokosing** [comment]
> is it possible to upgrade the Kafka stuff first, without upgrading OkHttp? 

That is great idea! Thanks

**Comment by kokosing** [comment]
> is it possible to upgrade the Kafka stuff first, without upgrading OkHttp? 

That is great idea! Thanks

**Comment by hashhar** [comment]
Partially superseded by https://github.com/trinodb/trino/pull/16763.

@wendigo Do you want to start fresh on this? Or should I close this?

**Comment by wendigo** [comment]
@hashhar I'll start fresh.
